### DISCUSSION TITLE
All-in-one middleware

### DISCUSSION THREAD

**Discussion by glenjamin** [klauspost/compress#845] [Ideas/Features]
Hello!

Would you accept a PR that adds a single server middleware which inspects the `Accept-Encoding` header and selects the best available compression implementation?

I could build this as a standalone project, but as there's already a bunch of different compression implementations here it seems like this might be a good fit?

I'm imagining something that decodes the header, and then selects `zstd`, `brotli` or `gzip` depending on what the client claims to support

**Comment by klauspost** [comment]
I haven't really pushed for it in gzhttp, since zstd isn't implemented in major browsers and brotli doesn't have Go compression implementation to know knowledge.

Technically the [Implementation](https://pkg.go.dev/github.com/klauspost/compress/gzhttp#Implementation) is pretty much implementation independent, except for padding. So adding a "method" to register other methods would be feasible.

Adding zstd could be done without too much effort I suspect.

Edit (didn't really answer your question): Yes, I'd be happy to have gzhttp do other types. Adding zstd by default would also be fine. If you can make a PR that doesn't break existing API.

If [BREACH mitigation](https://github.com/klauspost/compress/tree/master/gzhttp#breach-mitigation) is enabled, that should take priority, which means that gzip should be preferred until other methods can pad data.
